### PR TITLE
Docker/bazel docker rules

### PR DESCRIPTION
**Pull request by lopez-loic**
## Changes

- Rename server binary to "server_bin"

### Build the server bin
```bash
bazel-2.2.0 build --cxxopt=-std=c++17 //apps/app1:server_bin
```

### Server

```bash
bazel-2.2.0 run --cxxopt=-std=c++17 //apps/app1:server_image
```
***Output:***
```bash
loiclopez@loiclopez:/mnt/POC/test-pistache$ bazel-2.2.0 run --cxxopt=-std=c++17 //apps/app1:server_image
Starting local Bazel server and connecting to it...
INFO: Analyzed target //apps/app1:server_image (72 packages loaded, 6843 targets configured).
INFO: Found 1 target...
INFO: Deleting stale sandbox base /home/loiclopez/.cache/bazel/_bazel_loiclopez/d73b303449442f085f2583106fce5f56/sandbox
Target //apps/app1:server_image up-to-date:
  /home/loiclopez/.cache/bazel/_bazel_loiclopez/d73b303449442f085f2583106fce5f56/execroot/__main__/bazel-out/k8-fastbuild/bin/apps/app1/server_image-layer.tar
INFO: Elapsed time: 23.478s, Critical Path: 1.89s
INFO: 26 processes: 26 linux-sandbox.
INFO: Build completed successfully, 42 total actions
INFO: Build completed successfully, 42 total actions
Loaded image ID: sha256:cef3854686eabeedecc07975b525ce889d480a8ccd8584ecd5b548a7cb71752a
Tagging cef3854686eabeedecc07975b525ce889d480a8ccd8584ecd5b548a7cb71752a as bazel/apps/app1:server_image
```

#### Run the server container
```bash
docker run -d -p 9080:9080 bazel/apps/app1:server_image
```

### Client
```bash
bazel-2.2.0 run --cxxopt=-std=c++17 //apps/app2:client_image
```

***Output:***
```bash
loiclopez@loiclopez:/mnt/POC/test-pistache$ bazel-2.2.0 run --cxxopt=-std=c++17 //apps/app2:client_image
INFO: Analyzed target //apps/app2:client_image (1 packages loaded, 3 targets configured).
INFO: Found 1 target...
Target //apps/app2:client_image up-to-date:
  /home/loiclopez/.cache/bazel/_bazel_loiclopez/d73b303449442f085f2583106fce5f56/execroot/__main__/bazel-out/k8-fastbuild/bin/apps/app2/client_image-layer.tar
INFO: Elapsed time: 3.219s, Critical Path: 2.46s
INFO: 6 processes: 6 linux-sandbox.
INFO: Build completed successfully, 12 total actions
INFO: Build completed successfully, 12 total actions
Loaded image ID: sha256:4d892f0686756db5181e3057533967aab08ecbe3c5f5336b92c94899968f421b
Tagging 4d892f0686756db5181e3057533967aab08ecbe3c5f5336b92c94899968f421b as bazel/apps/app2:client_image
```

#### Run the client container
```bash
docker run --net=host --rm  bazel/apps/app2:client_image
```

### CODE REVIEW

File: apps/app1/BUILD [hunk 1]
```diff
@@ -1,9 +1,16 @@
...
+     deps = ["//third_party/getTimeOfDay:libtime"],
+     linkopts = [ "-lpistache" , "-lpthread"]
 )
 
+load("@io_bazel_rules_docker//container:container.bzl", "container_image")
 
-
+container_image(
+    name = "server_image",
+    base = "@cxx_container//image",
```
Comment (alexsaker) [inline apps/app1/BUILD:12]: Est-ce que le container marche sans que pistache soit installé dessus?

File: apps/app1/BUILD [hunk 2]
```diff
@@ -1,9 +1,16 @@
...
+    files = [":server_bin"],
+    entrypoint = ["./server_bin"],
+    ports = ["9080"]
+)
```
Comment (alexsaker) [inline apps/app1/BUILD:16]: J'aurais bien rajouté user, repository st(et optionellement: samp, labels,..)

File: apps/app1/BUILD (hunk 1 above)
Comment (lopez-loic) [reply apps/app1/BUILD:12]: Oui car par défaut la rule bazel `cc_binary` compile les librairies en mode statique : 

> linkstatic | Boolean; optional; default is True

**Comment by lopez-loic** [comment]
## New Changes

### Bazel docker_rules

- Downgrade to v0.11.0
- Downgrade bazel to v1.2.1

### Server
- Rename `server_image` to `v1`
- add `stamp = True`
- add `repository` rule parameter
- add `com.pistache.server` as image label

### Build the server image
```bash
bazel run --cxxopt=-std=c++17 //apps/app1:v1
```

***Output:***
```bash
loiclopez@loiclopez:/mnt/POC/test-pistache$ bazel run --cxxopt=-std=c++17 //apps/app1:v1
INFO: Invocation ID: f8d409c0-ed5a-4b8d-8878-c1c548aa5002
DEBUG: Rule 'cxx_container' indicated that a canonical reproducible form can be obtained by modifying arguments digest = "sha256:2e70e9c81838224b5311970dbf7ed16802fbfe19e7a70b3cbfa3d7522aa285b4"
DEBUG: Call stack for the definition of repository 'cxx_container' which is a container_pull (rule definition at /home/loiclopez/.cache/bazel/_bazel_loiclopez/d73b303449442f085f2583106fce5f56/external/io_bazel_rules_docker/container/pull.bzl:203:18):
 - /mnt/POC/test-pistache/WORKSPACE:45:1
INFO: Analyzed target //apps/app1:v1 (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //apps/app1:v1 up-to-date:
  /home/loiclopez/.cache/bazel/_bazel_loiclopez/d73b303449442f085f2583106fce5f56/execroot/test_pistache/bazel-out/k8-fastbuild/bin/apps/app1/v1-layer.tar
INFO: Elapsed time: 0.284s, Critical Path: 0.01s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
Loaded image ID: sha256:c393cc1da640d983fcaf1fbc9a401eab3d7cde77f9a1026311bde7c8d75211c7
Tagging c393cc1da640d983fcaf1fbc9a401eab3d7cde77f9a1026311bde7c8d75211c7 as pistache/apps/app1:v1
```

#### Show docker images

```bash
loiclopez@loiclopez:/mnt/POC/test-pistache$ docker images
REPOSITORY           TAG                 IMAGE ID            CREATED             SIZE
pistache/apps/app1   v1                  c393cc1da640        12 minutes ago      73.7MB
```

#### Run the server container
```bash
docker run -d -p 9080:9080 pistache/apps/app1:v1
```

### Client
```bash
bazel run --cxxopt=-std=c++17 //apps/app2:v1
```

***Output:***
```bash
loiclopez@loiclopez:/mnt/POC/test-pistache$ bazel run --cxxopt=-std=c++17 //apps/app2:v1
INFO: Invocation ID: 42e1b48d-173f-4e05-bcac-cae67182e486
DEBUG: Rule 'cxx_container' indicated that a canonical reproducible form can be obtained by modifying arguments digest = "sha256:2e70e9c81838224b5311970dbf7ed16802fbfe19e7a70b3cbfa3d7522aa285b4"
DEBUG: Call stack for the definition of repository 'cxx_container' which is a container_pull (rule definition at /home/loiclopez/.cache/bazel/_bazel_loiclopez/d73b303449442f085f2583106fce5f56/external/io_bazel_rules_docker/container/pull.bzl:203:18):
 - /mnt/POC/test-pistache/WORKSPACE:45:1
INFO: Analyzed target //apps/app2:v1 (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //apps/app2:v1 up-to-date:
  /home/loiclopez/.cache/bazel/_bazel_loiclopez/d73b303449442f085f2583106fce5f56/execroot/test_pistache/bazel-out/k8-fastbuild/bin/apps/app2/v1-layer.tar
INFO: Elapsed time: 0.249s, Critical Path: 0.01s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
Loaded image ID: sha256:a52e9e6172f3c136a9dea4791df7a97d6a6da6f58c95477c46a4ae2a0f014de8
Tagging a52e9e6172f3c136a9dea4791df7a97d6a6da6f58c95477c46a4ae2a0f014de8 as pistache/apps/app2:v1
```
#### Show docker images

```bash
loiclopez@loiclopez:/mnt/POC/test-pistache$ docker images
REPOSITORY           TAG                 IMAGE ID            CREATED             SIZE
pistache/apps/app2   v1                  a52e9e6172f3        12 minutes ago      74MB
pistache/apps/app1   v1                  c393cc1da640        12 minutes ago      73.7MB
```

#### Run the client container
```bash
docker run --net=host --rm  pistache/apps/app2:v1
```

File: apps/app1/BUILD (hunk 2 above)
Comment (lopez-loic) [reply apps/app1/BUILD:21]: Ajouts :
- repository
- stamp
- labels

**Review by dcorbin83 — APPROVED** [review]

**Pull request merged**